ddtrace/utils/cache.py

```diff
@@ -0,0 +1,81 @@
+from threading import RLock
+
+
+def cached(maxsize=256):
+    def cached_wrapper(f):
+        cache = {}
+        lock = RLock()
+        miss = object()
+
+        def cached_f(key):
+            if len(cache) >= maxsize:
+                for _, h in zip(range(maxsize >> 1), sorted(cache, key=lambda h: cache[h][1])):
+                    del cache[h]
+
+            _ = cache.get(key, miss)
+            if _ is not miss:
+                value, count = _
+                cache[key] = (value, count + 1)
+                return value
+
+            with lock:
+                _ = cache.get(key, miss)
+                if _ is not miss:
+                    value, count = _
+                    cache[key] = (value, count + 1)
+                    return value
+
+                result = f(key)
+
+                cache[key] = (result, 1)
+
+                return result
+
+        def invalidate(cache):
+            cache.clear()
+
+        cached_f.invalidate = lambda: invalidate(cache)
+
+        return cached_f
+
+    return cached_wrapper
+
+
+def cachedmethod(maxsize=256):
+    def cached_wrapper(f):
+        cache = {}
+        lock = RLock()
+        miss = object()
+
+        def cached_f(self, key):
+            if len(cache) >= maxsize:
+                for _, h in zip(range(maxsize >> 1), sorted(cache, key=lambda h: cache[h][1])):
+                    del cache[h]
+
+            _ = cache.get(key, miss)
+            if _ is not miss:
+                value, count = _
+                cache[key] = (value, count + 1)
+                return value
+
+            with lock:
+                _ = cache.get(key, miss)
+                if _ is not miss:
+                    value, count = _
+                    cache[key] = (value, count + 1)
+                    return value
+
+                result = f(self, key)
```

Comment:
What do you think about factoring out the common code in `cachedmethod` and `cached`?

Reply:
The main problem that I had to solve was to ensure that caching is not shared across instance methods. With that solved there isn't much common code to factor out, unless I start creating small utility functions that would bump the number of function calls

Reply:
In fact, with the introduction of a descriptor, I was able to reuse the code 🙂